### ISSUE TITLE
Blazor needs to be able to run as a PWA

### ISSUE THREAD

**Issue by xamlsharp** [aspnet/Blazor#183]
I'm extremely excited about this project, but even more so if Blazor could be integrated with the principles behind creating "Progressive Web Apps". Of course, this would mean that interop with the Service Worker would be required (among other things).

With the recent beta release of Google's Flutter, can Microsoft one-up them by using a better programming language with better tooling in order to create apps with an even greater reach? Blazor PWA's just might be the answer...

**Comment by SteveSandersonMS** [comment]
We're definitely keen to ensure Blazor can be used for PWAs! At the moment it seems like a fairly independent concept, in that you could use Blazor with PWA-type APIs to enable offline support and all the other PWA features, but you could also choose not to.

Eventually we might have some kind of out-of-the-box PWA template though for now we're focused on building the core technology.

**Comment by vanderstack** [comment]
@veikkoeeva are you able to recap any of what was mentioned or link to it or a timestamp? I missed the discussion, but am very interested in how we could use C# to build PWA's, and this idea being talked about within the aurelia community has my interest piqued even further.

**Comment by veikkoeeva** [comment]
@vanderstack It was more on private accounts, but there's something spilled publicly at https://gitter.im/aurelia/Discuss?at=5a8ea5c3888332ee3ab8c422. Contemplating more than sprining into action, as of now at least.

**Comment by deepness187** [comment]
Fuck you

**Comment by touseefbsb** [comment]
Definitely +1 to this, I've been ignoring web development for many years just bcz the chaos of too many frameworks. I am a c# developer and would love to see a full stack c# web development model to develop modren sin page web apps along with pwa features ( which is a must in all framework these days ) and if all of this can be done without touching any kind of vanilla js or even any js framework totally, it would be ideal. I can bear with basic html and css stuff for styling and UI, but no javascript please.

**Comment by rstropek** [comment]
@touseefbsb I am sharing your excitement about Blazor. However, I think for quite a long time working with Blazor will go hand in hand with working with JavaScript. There are so many useful JS libraries out there. It might take years (if it even ever happens) until all of them come with nice Blazor support. The same happened with TypeScript. It took quite a long time until JS libraries and TypeScript went along well. And TypeScript is nearer to JS than Blazor is.

I think every Blazor developer will need good HTML+CSS+__JavaScript/TypeScript__ skills in the near- to mid-term future.

**Comment by touseefbsb** [comment]
@rstropek  first think that typescript transpiles to javascript, but blazor runs .net directly into the web without converting into js, correct me if I am wrong though. but if this is true this changes everything. I understand we still need to learn html/css but to be honest that isnt as big a problem, the main issue was javascript if blazor can solve that, I am all in for web development. 
Also if blazor manager to create a proper template system like xaml style or any style near to it, then we wont be working much with html/css either, because then community will start to create components and controls for the ui using blazor and it will hide all the html/css from a normal user of the blazor controls. template controls based on porperties is the best thing in this scenario. but if u want advanced stuff or want to create/edit ur own controls then obviously u need html and css, but the third wheel can be and should be c# instead of js.

**Comment by fonix232** [comment]
@touseefbsb I doubt Blazor will drop the current Razor template base. It would be stupid to do so. The whole app runs on client side, and there are already pretty well written HTML+CSS renderers, why would the team want to write their own?

Sticking to the current Razor templates is the way in my opinion.

Also, for JS... Most of the web is in JS. Even Bootstrap uses quite some JS logic to do stuff. Right now you have two options: rewrite all frameworks with Blazor and C#, or write simple "binding libraries" for such components. I think in the near future, Vue/React components will be easily ported, similar to how the Xamarin project wrote Objective Sharpie for generating binding libraries. And I think the latter will prevail. But until that is done, you'll have to write JS code.

**Comment by touseefbsb** [comment]
then what is the main advantage of blazor if we still have to write js? @fonix232

**Comment by zbecknell** [comment]
@touseefbsb I don’t think anyone is saying you *have* to write JS, but that you *should* remain familiar as a practicality for the foreseeable future.  I’m with you on the desire to write less JS, but it will take time.

**Comment by touseefbsb** [comment]
@zbecknell  yeah I understand it will take time, but atleast its a good start with blazor :)

**Comment by Jack85** [comment]
Vue reactivity, component communication and integration with blazor, that would be awesome!

**Comment by touseefbsb** [comment]
the very basic reason of blazor was so that we dnt have to use frameworks like vue and angular etc, I really hope blazor sticks to its own framework and keep making it better for c# developers, instead of integrating it with other popular frameworks, which will increase the problem it was supposed to solve :)

**Comment by touseefbsb** [comment]
@danielmeza  all these wrappers just make me uncomfortable bcz then this will be another xamarin, which can never be 100 percent native performant, yes writing the whole framework can take time but it will be worth it to start from root and build it all up in c#instead of just wraping everything we dnt wanna use and make some parts of app in c#

**Comment by fonix232** [comment]
Why would it be odd? It would be a merge of the best of both worlds - the simplicity and syntax of C# while using the existing structure and objects of Vue/React/etc. 

Yes, it doe not alleviate the need for a native framework, but with the amount of resources available written in JS, it would be stupid to not have an easy option to use them.

**Comment by nicowernli** [comment]
@fonix232 I don't see the need of Vue Components or Javascript libraries if you can write the code in C# with Blazor. Actually what Blazor does is what Vue and React do but better, because it doesn't use Javascript (btw I'm a full-time JS developer). Thats the whole point of WebAssembly, using any other programming language apart from Javascript. If you are going to implement Vue or React in Blazor whats the point? You better keep writting Vue or React apps without Blazor and thats it.

**Comment by fonix232** [comment]
@nicowernli you might not see it, but someone coming from an existing knowledgebase of Vue and JS libraries who does not want to rewrite the 200+ dependencies a project would have, that person sees the importance of converting step by step.

First step is having the overall final framework set as Blazor, that uses a few JS libraries, Vue/React, and so on. So they put their existing codebase in a wrapper for Blazor, and first replace smaller libraries with optimized C# code, then step by step phase out the rest of the elements.

This would also help with initial adoption - sure, it will be JS running for some time, but until there's a Blazor component library at least on par with React's baseline offering, I don't see many people converting.

**Comment by nicowernli** [comment]
I don't think people that is ok using JS will "convert". And after using Blazor in really small things (I'm still learning it) I can't find many differences against Nuxt, or Next to be honest.
If you already have a massive app, migrating it anything different that what is using will be a pain in the **se anyway.

**Comment by danielmeza** [comment]
@nicowernli  and @touseefbsb  I'm agree with you both and the goal for this proyect must be "Create a C# way to work on web client" but some nesesary functionality are missing yet, lke SignalR until they can be impelmented. 
We need to mantain the gap there to avoid covert the framework in other Xamarin like @nicowernli  say

**Comment by danroth27** [comment]
> but some nesesary functionality are missing yet, lke SignalR

There is now a Blazor SignalR client being maintained by the community: https://github.com/BlazorExtensions/SignalR

**Comment by touseefbsb** [comment]
thankyou for the reference @danroth27 , u just proved how fast community is willing to help once we get serious abt this, people really want to get as far from JS as possible, as u can see JS developers like @nicowernli are also considering blazor, the thing is I understand interop might help in migration etc, but if interop exists, eventually blazor will become lazy project bcz u can do most of the stuff with JS interops and blazor itself will never become an original framework without js interop dependency, or it may take too long to become one.
on the other hand if our complete focus is making it 100 percent pure c# framework then whole community of .net developers will be willing to help and its progress will be really fast and very soon it will be as good as Vue, react, angular, and then u will see it will be doing everything they can and some stuff which even vue, react cannot, because web assembly can do everything better in a more simpler way, and interops will just make blazor too lazy to re-invent the wheel in c# for its own good.

**Comment by Terricide** [comment]
If you are a UWP developer you should check out the uno platform.

**Comment by touseefbsb** [comment]
yes I have checked the uno platform but it is still in early days and the product still has its flaws and very less support in the community but it is definitely something I am hopeful about in the future :) @Terricide

**Comment by krdmllr** [comment]
I'm currently working on creating a pwa using blazor. I just added all files that are required for a pwa to a blazor project, you can find the pwa here: https://krdmllr.github.io/DotnetPwaSample/ and the code here: https://github.com/krdmllr/DotnetPwaSample
Would the added serviceworker/manifest and modified index.html be enough for the template or would this require some data caching strategy aswell?
Also, the serviceworker is currently written in js and you need to add all .dll's to a list. I could not find a way to auto generate the list or work with wasm here yet.

**Comment by danroth27** [comment]
> Would the added serviceworker/manifest and modified index.html be enough for the template or would this require some data caching strategy as well?

I was thinking we would want to provide some prebuilt caching strategies, similar to what @madskristensen worked on for ASP.NET Core a while back: https://madskristensen.net/blog/progressive-web-apps-made-easy-with-aspnet-core/

> Also, the serviceworker is currently written in js and you need to add all .dll's to a list. I could not find a way to auto generate the list or work with wasm here yet.

At some point we may want to look at implementing service workers in .NET, but for now I think implementing them in JS is ok. For the list of assemblies I think you might be able to leverage the `blazer.boot.json` file that get produced when building the app.

**Comment by krdmllr** [comment]
@danroth27 thank you for the infos. The blog post is interesting, i will see if i can use some of the ideas in my pwa. mads is using a backend to generate some of the files, i thought a standalone client project would be better for a template, what do you think?

**Comment by danroth27** [comment]
> i thought a standalone client project would be better for a template, what do you think?

Yup, I agree.

**Comment by danielterabyte** [comment]
Hi, guys.
Is there any update in this matter?

I am starting a POC at work and it would be very good to have PWA Blazor app.

Thanks.

**Comment by danroth27** [comment]
No progress on our side, but there are some community efforts based on blazor you might look at, like [Bionic](https://bionicframework.github.io/)

**Comment by richbryant** [comment]
Dead link there, @danroth27

**Comment by treg** [comment]
https://github.com/BionicFramework/
